### PR TITLE
Update :-

### PR DESCRIPTION
**Pull request by littlefermatlives**
1. Add run template query endpoint in routes.go
2. Add run template query request handler in handlers.go
3. Add run template query task
4. Add task result for query task
5. Add SDK exceptions for qureyfailureexceptions

### CODE REVIEW

File: cmd/api/handler.go [hunk 1]
```diff
@@ -505,3 +505,39 @@ func (app *Config) singleReplaceTask(w http.ResponseWriter, r *http.Request) {
...
 	_ = app.writeJSON(w, http.StatusOK, resPayload)
 }
+
+func (app *Config) runQueryTask(w http.ResponseWriter, r *http.Request) {
+	task := &tasks.QueryTask{}
+	if err := app.readJSON(w, r, task); err != nil {
+		_ = app.errorJSON(w, err, http.StatusUnprocessableEntity)
+		return
+	}
+	log.Print(task, "Run Query")
```
Comment (manik-couchbase) [inline cmd/api/handler.go:515]: log.Println(task, tasks.QueryOperation)

**Comment by manik-couchbase** [comment]
I think task_query is missing in /internal/tasks/....

File: cmd/api/handler.go (hunk 1 above)
Comment (littlefermatlives) [reply cmd/api/handler.go]: updated

File: cmd/api/handler.go [hunk 2]
```diff
@@ -505,3 +505,39 @@ func (app *Config) singleReplaceTask(w http.ResponseWriter, r *http.Request) {
 	}
 	_ = app.writeJSON(w, http.StatusOK, resPayload)
 }
+
```
Comment (manik-couchbase) [inline cmd/api/handler.go:508]: Add comment to describe function

File: internal/task_result/task_result.go [hunk 3]
```diff
@@ -26,6 +26,11 @@ type SingleOperationResult struct {
 	Cas         uint64 `json:"cas"  doc:"true"`
 }
 
+type FailedQuery struct {
```
Comment (manik-couchbase) [inline internal/task_result/task_result.go:29]: Add doc json tag for documentation.
Add the structs with correct naming at  https://github.com/couchbaselabs/sirius/blob/6fdcc5cbd2ca3a2b793cdb80d47efd63dff58d29/internal/generate/register.go#L36

File: internal/tasks/helper.go [hunk 4]
```diff
@@ -89,6 +91,24 @@ func configSingleOperationConfig(s *SingleOperationConfig) error {
 	return nil
 }
 
+type QueryOperationConfig struct {
+	template         string `json:"template,omitempty" doc:"true"`
```
Comment (manik-couchbase) [inline internal/tasks/helper.go:95]: make template string a public var

File: internal/tasks/helper.go [hunk 5]
```diff
@@ -89,6 +91,24 @@ func configSingleOperationConfig(s *SingleOperationConfig) error {
...
+	BuildIndexViaSDK bool   `json:"buildIndexViaSDK" doc:"true"`
+}
+
+func configQueryOperationConfig(s *QueryOperationConfig) error {
+	if s == nil {
+		return fmt.Errorf("unable to parse QueryOperationConfig")
+	}
+
+	if s.Duration == 0 || s.Duration > MaxQueryRuntime {
+		s.Duration = 100
```
Comment (manik-couchbase) [inline internal/tasks/helper.go:107]: make this 100 as a constant and can be declared as DefaultQueryRunTime

File: internal/tasks/task_run_query.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+		log.Println(task.result.ErrorOther)
+		if err := task.result.SaveResultIntoFile(); err != nil {
+			log.Println("not able to save result into ", task.ResultSeed)
+			return err
+		}
+		return task.tearUp()
+	} else {
+		task.result = task_result.ConfigTaskResult(task.Operation, task.ResultSeed)
+	}
+	cluster, err := task.req.connectionManager.GetCluster(task.ClusterConfig)
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:112]: Add a empty line here for better readability

File: internal/tasks/task_run_query.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+	task.result.Success = task.result.Success - task.result.Failure
+	log.Println(fmt.Sprintf("number of success query: %d", task.result.Success))
+	if err := task.result.SaveResultIntoFile(); err != nil {
+		log.Println("not able to save result into ", task.ResultSeed)
+	}
+	return task.tearUp()
+}
+
+func buildIndexSDKManager(cluster *gocb.Cluster, bucketName string, scopeName string, collectionName string, Type string, indexName string, fieldNameList []string) error {
+	manager := cluster.QueryIndexes()
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:161]: Can be remaned as queryManager

File: internal/tasks/task_run_query.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+	log.Println(fmt.Sprintf("number of failed query: %d", task.result.Failure))
+	task.result.Success = task.result.Success - task.result.Failure
+	log.Println(fmt.Sprintf("number of success query: %d", task.result.Success))
+	if err := task.result.SaveResultIntoFile(); err != nil {
+		log.Println("not able to save result into ", task.ResultSeed)
+	}
+	return task.tearUp()
+}
+
+func buildIndexSDKManager(cluster *gocb.Cluster, bucketName string, scopeName string, collectionName string, Type string, indexName string, fieldNameList []string) error {
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:160]: Type as IndexType

File: internal/tasks/task_run_query.go [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+	manager := cluster.QueryIndexes()
+	if Type == "create primary index" {
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:162]: "create primary index", "create index", "build index" declare them as constants

File: internal/tasks/task_run_query.go [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+			&gocb.CreateQueryIndexOptions{Deferred: true, ScopeName: scopeName, CollectionName: collectionName},
+		); err != nil {
+			return err
+		}
+	} else if Type == "build index" {
+		indexesToBuild, err := manager.BuildDeferredIndexes(bucketName, nil)
+		if err != nil {
+			return err
+		}
+		err = manager.WatchIndexes(bucketName, indexesToBuild, time.Duration(120*time.Second), nil)
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:179]: Make 120 as constant, can be WatchIndexesDuration

File: internal/tasks/task_run_query.go [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+func buildIndexWithSDKs(task *QueryTask, cluster *gocb.Cluster, scope *gocb.Scope, collection *gocb.Collection) {
+	//Create Primary index
+	log.Println("Building index to run query with sdk")
+	err := buildIndexSDKManager(cluster, task.Bucket, scope.Name(), collection.Name(), "create primary index", "", []string{""})
+	if err != nil {
+		log.Println("build primary index failed")
+		task.result.IncrementQueryFailure("Create primary index query", err)
+	}
+	//Create index on firstName, lastName, attributes, address
+	err = buildIndexSDKManager(cluster, task.Bucket, scope.Name(), collection.Name(),
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:196]: This firstName, lastName is associated with a specific person template.
We have to make it specific template agnostic.

Let's say template.GetIndexFields() -> returns([]string , nameOfindex) which can be used to build indexes.

Comment (manik-couchbase) [reply internal/tasks/task_run_query.go:196]: Put them in a loop,

for different_index_names, []feilds of that indexes -> create index

**Review by manik-couchbase — CHANGES_REQUESTED** [review]

File: cmd/api/handler.go (hunk 2 above)
Comment (littlefermatlives) [reply cmd/api/handler.go:508]: added comment

File: internal/task_result/task_result.go (hunk 3 above)
Comment (littlefermatlives) [reply internal/task_result/task_result.go:29]: added

File: internal/tasks/helper.go (hunk 4 above)
Comment (littlefermatlives) [reply internal/tasks/helper.go:95]: done

File: internal/tasks/helper.go (hunk 5 above)
Comment (littlefermatlives) [reply internal/tasks/helper.go:107]: done

File: internal/tasks/task_run_query.go (hunk 8 above)
Comment (littlefermatlives) [reply internal/tasks/task_run_query.go:160]: done

File: internal/tasks/task_run_query.go (hunk 9 above)
Comment (littlefermatlives) [reply internal/tasks/task_run_query.go:162]: done

File: internal/tasks/task_run_query.go (hunk 10 above)
Comment (littlefermatlives) [reply internal/tasks/task_run_query.go:179]: done

File: internal/tasks/task_run_query.go (hunk 11 above)
Comment (littlefermatlives) [reply internal/tasks/task_run_query.go:196]: done

File: internal/tasks/task_run_query.go [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+func buildIndexWithSDKs(task *QueryTask, cluster *gocb.Cluster, scope *gocb.Scope, collection *gocb.Collection) {
+	//Create Primary index
+	err := buildIndexSDKManager(cluster, task.Bucket, scope.Name(), collection.Name(), "createPrimaryIndex", "", []string{""})
+	if err != nil {
+		task.result.IncrementQueryFailure(fmt.Sprintf("Create primary index on `%s`.%s.%s",
+			task.Bucket, scope.Name(), collection.Name()), err)
+	}
+	// create indexes
+	indexes, err := task.gen.Template.GenerateIndexesForSdk()
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:198]: if indexes, err := task.gen.Template.GenerateIndexesForSdk(); err ! nil {
}

File: internal/tasks/task_run_query.go [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+	if err != nil {
+		log.Println("Get indexes for sdk failed.....")
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:200]: return

File: internal/tasks/task_run_query.go [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+	}
+	for indexName, indexFields := range indexes {
+		err = buildIndexSDKManager(cluster, task.Bucket, scope.Name(), collection.Name(),
+			"createIndex", indexName, indexFields)
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:204]: use the constant which is declared

File: internal/tasks/task_run_query.go [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+	if err != nil {
+		log.Println("Get sample queries failed.....")
+	}
+	// run a loop for a requested duration
+	expirationTime := time.Now().Add(time.Duration(task.QueryOperationConfig.Duration) * time.Second)
+	for time.Now().Before(expirationTime) {
+		routineLimiter <- struct{}{}
+		group.Go(func() error {
+			for i := 0; i < len(queries); i++ {
+				rows, err := cluster.Query(queries[i], &gocb.QueryOptions{})
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:256]: if ; {
}

File: internal/tasks/task_run_query.go [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+	result               *task_result.TaskResult      `json:"-" doc:"false"`
+	gen                  *docgenerator.QueryGenerator `json:"-" doc:"false"`
+	req                  *Request                     `json:"-" doc:"false"`
+}
+
+func (task *QueryTask) Describe() string {
+	return " Query task runs N1QL query over a period of time over a bucket.\n"
+}
+
+func (task *QueryTask) BuildIdentifier() string {
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:36]: Build Identifier cross check with task_insert
reference - https://github.com/couchbaselabs/sirius/blob/main/internal/tasks/task_bulk_insert.go#L46C1-L51C2

File: internal/tasks/task_run_query.go [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+	}
+	return fmt.Sprintf("%s-%s-%s-%s-%s", task.ClusterConfig.Username, task.IdentifierToken, task.Bucket, task.Scope,
+		task.Collection)
+}
+
+func (task *QueryTask) CheckIfPending() bool {
+	return task.TaskPending
+}
+
+func (task *QueryTask) Config(req *Request, seed int, seedEnd int, reRun bool) (int, error) {
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:58]: cross check with new Config() implementation 
Reference - https://github.com/couchbaselabs/sirius/blob/main/internal/tasks/task_bulk_insert.go#L58

**Review by manik-couchbase — CHANGES_REQUESTED** [review]

File: internal/tasks/task_run_query.go [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+			task.Bucket = DefaultBucket
+		}
+		if task.Scope == "" {
+			task.Scope = DefaultScope
+		}
+		if task.Collection == "" {
+			task.Collection = DefaultCollection
+		}
+
+		if task.IdentifierToken == "" {
```
Comment (manik-couchbase) [inline internal/tasks/task_run_query.go:79]: remove IdentifierToken from here, We give it default value in BuildIdentifier()

**Review by manik-couchbase — CHANGES_REQUESTED** [review]

File: internal/tasks/task_run_query.go (hunk 18 above)
Comment (littlefermatlives) [reply internal/tasks/task_run_query.go]: removed

**Review by manik-couchbase — APPROVED** [review]

**Pull request merged**